File: next/src/app/toolRegistry.ts
import {
  ArrowRightLeft,
  Binary,
  Braces,
  Calculator,
  CalendarClock,
  Clock3,
  CodeXml,
  Columns2,
  Cpu,
  FileCog,
  FileText,
  Globe2,
  Home,
  Image,
  Languages,
  LockKeyhole,
  Network,
  NotebookPen,
  Palette,
  Paintbrush,
  QrCode,
  Regex,
  ServerCog,
  SquareTerminal,
  Variable,
  type LucideIcon
} from 'lucide-react'
import { lazy, type ComponentType, type LazyExoticComponent } from 'react'
import type { MessageKey } from '@/shared/i18n/messages'

export const toolIds = [
  'mootool',
  'quickNote',
  'textDiff',
  'reformat',
  'json',
  'java',
  'ymlProperties',
  'protobuf',
  'variables',
  'http',
  'host',
  'net',
  'uaParse',
  'encode',
  'crypto',
  'regex',
  'cron',
  'qrCode',
  'timeConvert',
  'translation',
  'calculator',
  'colorBoard',
  'image',
  'pdf',
  'hardware'
] as const

export type ToolId = (typeof toolIds)[number]
export type ToolGroupId = 'text' | 'dev' | 'network' | 'encode' | 'daily' | 'system'
export type ToolStatus = 'placeholder' | 'in-progress' | 'parity-review' | 'complete'

export type ToolDefinition = {
  id: ToolId
  groupId: ToolGroupId | 'home'
  titleKey: MessageKey
  keywords: string[]
  icon: LucideIcon
  component?: LazyExoticComponent<ComponentType>
  status: ToolStatus
  supportsHistory: boolean
  supportsFavorites: boolean
  settingsSection?: string
}

export type ToolGroupDefinition = {
  id: ToolGroupId
  titleKey: MessageKey
  toolIds: ToolId[]
}

const HomePage = lazy(() => import('@/features/home/HomePage').then((module) => ({ default: module.HomePage })))
const JsonTool = lazy(() => import('@/features/json/JsonTool').then((module) => ({ default: module.JsonTool })))
const TimeConvertTool = lazy(() => import('@/features/time/TimeConvertTool').then((module) => ({ default: module.TimeConvertTool })))
const EncodeTool = lazy(() => import('@/features/encode/EncodeTool').then((module) => ({ default: module.EncodeTool })))
const UaParseTool = lazy(() => import('@/features/ua/UaParseTool').then((module) => ({ default: module.UaParseTool })))
const CalculatorTool = lazy(() => import('@/features/calculator/CalculatorTool').then((module) => ({ default: module.CalculatorTool })))
const RegexTool = lazy(() => import('@/features/regex/RegexTool').then((module) => ({ default: module.RegexTool })))
const CronTool = lazy(() => import('@/features/cron/CronTool').then((module) => ({ default: module.CronTool })))
const TextDiffTool = lazy(() => import('@/features/diff/TextDiffTool').then((module) => ({ default: module.TextDiffTool })))
const ConfigConvertTool = lazy(() => import('@/features/config/ConfigConvertTool').then((module) => ({ default: module.ConfigConvertTool })))
const ReformatTool = lazy(() => import('@/features/reformat/ReformatTool').then((module) => ({ default: module.ReformatTool })))
const ProtobufTool = lazy(() => import('@/features/protobuf/ProtobufTool').then((module) => ({ default: module.ProtobufTool })))
const CryptoTool = lazy(() => import('@/features/crypto/CryptoTool').then((module) => ({ default: module.CryptoTool })))
const QrCodeTool = lazy(() => import('@/features/qrcode/QrCodeTool').then((module) => ({ default: module.QrCodeTool })))
const ColorBoardTool = lazy(() => import('@/features/color/ColorBoardTool').then((module) => ({ default: module.ColorBoardTool })))
const ImageTool = lazy(() => import('@/features/image/ImageTool').then((module) => ({ default: module.ImageTool })))
const PdfTool = lazy(() => import('@/features/pdf/PdfTool').then((module) => ({ default: module.PdfTool })))
const VariablesTool = lazy(() => import('@/features/variables/VariablesTool').then((module) => ({ default: module.VariablesTool })))
const HttpTool = lazy(() => import('@/features/http/HttpTool').then((module) => ({ default: module.HttpTool })))
const HostTool = lazy(() => import('@/features/host/HostTool').then((module) => ({ default: module.HostTool })))
const NetTool = lazy(() => import('@/features/net/NetTool').then((module) => ({ default: module.NetTool })))
const TranslationTool = lazy(() => import('@/features/translation/TranslationTool').then((module) => ({ default: module.TranslationTool })))
const HardwareTool = lazy(() => import('@/features/hardware/HardwareTool').then((module) => ({ default: module.HardwareTool })))
const QuickNoteTool = lazy(() => import('@/features/quickNote/QuickNoteTool').then((module) => ({ default: module.QuickNoteTool })))
const RuntimeTool = lazy(() => import('@/features/runtime/RuntimeTool').then((module) => ({ default: module.RuntimeTool })))

export const toolRegistry: ToolDefinition[] = [
  tool('mootool', 'home', 'app.nav.home', Home, ['home', 'about', '首页', '主页', '关于'], HomePage, 'in-progress'),
  tool('quickNote', 'text', 'app.nav.quickNote', NotebookPen, ['note', 'memo', 'markdown', '随手记', '笔记', '记事'], QuickNoteTool, 'in-progress'),
  tool('textDiff', 'text', 'app.nav.textDiff', Columns2, ['diff', 'compare', 'text', '对比', '比较'], TextDiffTool, 'parity-review'),
  tool('reformat', 'text', 'app.nav.reformat', Paintbrush, ['format', 'java', 'xml', 'html', 'nginx', '格式化'], ReformatTool, 'parity-review'),
  tool('json', 'dev', 'app.nav.json', Braces, ['json', 'xml', 'javabean', 'jsonpath', '格式化'], JsonTool, 'in-progress'),
  tool('java', 'dev', 'app.nav.java', SquareTerminal, ['java', 'groovy', 'python', 'node', 'console', '运行'], RuntimeTool, 'in-progress'),
  tool('ymlProperties', 'dev', 'app.nav.ymlProperties', FileCog, ['yaml', 'yml', 'properties', 'config', '配置', '转换'], ConfigConvertTool, 'parity-review'),
  tool('protobuf', 'dev', 'app.nav.protobuf', Binary, ['protobuf', 'proto', 'wire', '序列化'], ProtobufTool, 'parity-review'),
  tool('variables', 'dev', 'app.nav.variables', Variable, ['env', 'environment', 'variable', '环境变量'], VariablesTool, 'parity-review'),
  tool('http', 'network', 'app.nav.http', Globe2, ['http', 'curl', 'api', 'request', '请求'], HttpTool, 'parity-review'),
  tool('host', 'network', 'app.nav.host', ServerCog, ['host', 'dns', '域名'], HostTool, 'parity-review'),
  tool('net', 'network', 'app.nav.net', Network, ['network', 'ip', 'ping', 'whois', '网络'], NetTool, 'parity-review'),
  tool('uaParse', 'network', 'app.nav.uaParse', CodeXml, ['ua', 'user-agent', 'browser', '浏览器', '分析'], UaParseTool, 'parity-review'),
  tool('encode', 'encode', 'app.nav.encode', ArrowRightLeft, ['encode', 'decode', 'base64', 'url', '编码', '解码'], EncodeTool, 'parity-review'),
  tool('crypto', 'encode', 'app.nav.crypto', LockKeyhole, ['crypto', 'hash', 'md5', 'sha', 'random', '加密', '随机'], CryptoTool, 'parity-review'),
  tool('regex', 'encode', 'app.nav.regex', Regex, ['regex', 'regexp', 'regular', '正则', '匹配'], RegexTool, 'parity-review'),
  tool('cron', 'encode', 'app.nav.cron', CalendarClock, ['cron', 'schedule', '定时', '表达式'], CronTool, 'parity-review'),
  tool('qrCode', 'encode', 'app.nav.qrCode', QrCode, ['qr', 'qrcode', '二维码', '条码'], QrCodeTool, 'parity-review'),
  tool('timeConvert', 'daily', 'app.nav.timeConvert', Clock3, ['time', 'timestamp', 'clock', '时间', '时区', '时间戳'], TimeConvertTool, 'parity-review'),
  tool('translation', 'daily', 'app.nav.translation', Languages, ['translate', 'translation', 'word', '翻译', '单词'], TranslationTool, 'parity-review'),
  tool('calculator', 'daily', 'app.nav.calculator', Calculator, ['calculator', 'calc', 'math', '计算', '表达式'], CalculatorTool, 'parity-review'),
  tool('colorBoard', 'daily', 'app.nav.colorBoard', Palette, ['color', 'palette', 'hex', 'rgb', '调色', '颜色'], ColorBoardTool, 'parity-review'),
  tool('image', 'daily', 'app.nav.image', Image, ['image', 'watermark', 'compress', '图片', '图像'], ImageTool, 'parity-review'),
  tool('pdf', 'daily', 'app.nav.pdf', FileText, ['pdf', 'merge', 'split', '合并', '拆分'], PdfTool, 'parity-review'),
  tool('hardware', 'system', 'app.nav.hardware', Cpu, ['hardware', 'system', 'cpu', 'memory', '系统', '硬件'], HardwareTool, 'parity-review')
]

export const toolGroups: ToolGroupDefinition[] = [
  group('text', 'app.group.text'),
  group('dev', 'app.group.dev'),
  group('network', 'app.group.network'),
  group('encode', 'app.group.encode'),
  group('daily', 'app.group.daily'),
  group('system', 'app.group.system')
]

export const toolById = new Map(toolRegistry.map((definition) => [definition.id, definition]))

export function isToolId(value: string): value is ToolId {
  return toolById.has(value as ToolId)
}

function tool(
  id: ToolId,
  groupId: ToolDefinition['groupId'],
  titleKey: MessageKey,
  icon: LucideIcon,
  keywords: string[],
  component?: LazyExoticComponent<ComponentType>,
  status: ToolStatus = 'placeholder'
): ToolDefinition {
  return {
    id,
    groupId,
    titleKey,
    keywords,
    icon,
    component,
    status,
    supportsHistory: id !== 'mootool' && id !== 'hardware',
    supportsFavorites: ['regex', 'cron', 'colorBoard'].includes(id)
  }
}

function group(id: ToolGroupId, titleKey: MessageKey): ToolGroupDefinition {
  return {
    id,
    titleKey,
    toolIds: toolRegistry.filter((definition) => definition.groupId === id).map((definition) => definition.id)
  }
}

File: next-compose/composeApp/src/desktopMain/kotlin/com/rememberber/mootool/next/compose/app/ToolRegistry.kt
package com.rememberber.mootool.next.compose.app

import com.rememberber.mootool.next.compose.model.ToolGroupId
import com.rememberber.mootool.next.compose.model.ToolId
import com.rememberber.mootool.next.compose.model.ToolStatus
import java.util.Locale

data class ToolDefinition(
    val id: ToolId,
    val groupId: ToolGroupId,
    val titleKey: String,
    val keywords: List<String>,
    val glyph: String,
    val status: ToolStatus,
    val supportsHistory: Boolean,
    val supportsFavorites: Boolean
)

data class ToolGroupDefinition(
    val id: ToolGroupId,
    val titleKey: String,
    val toolIds: List<ToolId>
)

object ToolRegistry {
    val tools: List<ToolDefinition> = listOf(
        tool(ToolId.Mootool, ToolGroupId.Home, "app.nav.home", "⌂", listOf("home", "about", "contributor", "sponsor", "首页", "主页", "贡献者", "ホーム")),
        tool(ToolId.QuickNote, ToolGroupId.Text, "app.nav.quickNote", "✎", listOf("note", "memo", "markdown", "随手记", "笔记"), ToolStatus.Available),
        tool(ToolId.TextDiff, ToolGroupId.Text, "app.nav.textDiff", "⇄", listOf("diff", "compare", "whitespace", "history", "对比"), ToolStatus.Available),
        tool(ToolId.Reformat, ToolGroupId.Text, "app.nav.reformat", "✦", listOf("format", "java", "xml", "html", "nginx", "prettier", "history", "格式化"), ToolStatus.Available),
        tool(
            ToolId.Json,
            ToolGroupId.Dev,
            "app.nav.json",
            "{}",
            listOf("json", "xml", "javabean", "jsonpath", "vault", "snippet", "文档库", "格式化"),
            ToolStatus.Available,
        ),
        tool(
            ToolId.Java,
            ToolGroupId.Dev,
            "app.nav.java",
            ">_",
            listOf("java", "groovy", "python", "node", "runtime", "detect", "codrun", "运行"),
            ToolStatus.Available,
        ),
        tool(ToolId.YmlProperties, ToolGroupId.Dev, "app.nav.ymlProperties", "☰", listOf("yaml", "yml", "properties", "snakeyaml", "配置"), ToolStatus.Available),
        tool(ToolId.Protobuf, ToolGroupId.Dev, "app.nav.protobuf", "⬡", listOf("protobuf", "proto", "wire", "hex", "base64"), ToolStatus.Available),
        tool(ToolId.Variables, ToolGroupId.Dev, "app.nav.variables", "∑", listOf("env", "environment", "环境变量"), ToolStatus.Available),
        tool(ToolId.Http, ToolGroupId.Network, "app.nav.http", "⇄", listOf("http", "curl", "api", "请求"), ToolStatus.Available),
        tool(ToolId.Host, ToolGroupId.Network, "app.nav.host", "◎", listOf("host", "hosts", "profile", "dns", "域名", "方案"), ToolStatus.Available),
        tool(ToolId.Net, ToolGroupId.Network, "app.nav.net", "◈", listOf("network", "ip", "ping", "网络"), ToolStatus.Available),
        tool(ToolId.UaParse, ToolGroupId.Network, "app.nav.uaParse", "UA", listOf("ua", "user-agent", "browser", "chrome", "浏览器"), ToolStatus.Available),
        tool(ToolId.Encode, ToolGroupId.Encode, "app.nav.encode", "⌁", listOf("encode", "decode", "base64", "编码"), ToolStatus.Available),
        tool(ToolId.Crypto, ToolGroupId.Encode, "app.nav.crypto", "🔒", listOf("crypto", "hash", "md5", "加密", "aes", "rsa"), ToolStatus.Available),
        tool(ToolId.Regex, ToolGroupId.Encode, "app.nav.regex", ".*", listOf("regex", "regular", "pattern", "正则", "捕获"), ToolStatus.Available),
        tool(ToolId.Cron, ToolGroupId.Encode, "app.nav.cron", "⏱", listOf("cron", "schedule", "quartz", "favorite", "history", "定时"), ToolStatus.Available),
        tool(ToolId.QrCode, ToolGroupId.Encode, "app.nav.qrCode", "▦", listOf("qr", "qrcode", "二维码"), ToolStatus.Available),
        tool(ToolId.TimeConvert, ToolGroupId.Daily, "app.nav.timeConvert", "◷", listOf("time", "timestamp", "时间"), ToolStatus.Available),
        tool(ToolId.MessageBoard, ToolGroupId.Daily, "app.nav.messageBoard", "✉", listOf("message", "board", "留言板", "留言", "沉浸"), ToolStatus.Available),
        tool(ToolId.Translation, ToolGroupId.Daily, "app.nav.translation", "文", listOf("translate", "word", "翻译"), ToolStatus.Available),
        tool(ToolId.Calculator, ToolGroupId.Daily, "app.nav.calculator", "=", listOf("calculator", "calc", "gcd", "lcm", "计算", "进制"), ToolStatus.Available),
        tool(ToolId.ColorBoard, ToolGroupId.Daily, "app.nav.colorBoard", "◐", listOf("color", "palette", "hex", "rgb", "调色", "取色"), ToolStatus.Available),
        tool(ToolId.Image, ToolGroupId.Daily, "app.nav.image", "▣", listOf("image", "图片", "svg", "watermark", "compress", "截图"), ToolStatus.Available),
        tool(ToolId.Pdf, ToolGroupId.Daily, "app.nav.pdf", "▤", listOf("pdf", "merge", "split", "拆分", "合并"), ToolStatus.Available),
        tool(ToolId.Hardware, ToolGroupId.System, "app.nav.hardware", "⚙", listOf("hardware", "system", "cpu", "系统", "oshi"), ToolStatus.Available)
    )

    val groups: List<ToolGroupDefinition> = listOf(
        ToolGroupId.Text, ToolGroupId.Dev, ToolGroupId.Network, ToolGroupId.Encode, ToolGroupId.Daily, ToolGroupId.System
    ).map { group ->
        ToolGroupDefinition(group, "app.group.${group.id}", tools.filter { it.groupId == group }.map { it.id })
    }

    val byId: Map<ToolId, ToolDefinition> = tools.associateBy { it.id }

    fun search(query: String, translate: (String) -> String): List<ToolDefinition> {
        val needle = query.trim().lowercase(Locale.ROOT)
        if (needle.isEmpty()) return tools
        return tools.filter { tool -> matchesSearch(needle, tool, translate) }
    }

    /** Locale-stable id/keyword matching; localized titles use default locale (Electron `toLocaleLowerCase`). */
    internal fun matchesSearch(needle: String, tool: ToolDefinition, translate: (String) -> String): Boolean {
        if (tool.id.id.lowercase(Locale.ROOT).contains(needle)) return true
        if (translate(tool.titleKey).lowercase(Locale.getDefault()).contains(needle)) return true
        return tool.keywords.any { it.lowercase(Locale.ROOT).contains(needle) }
    }

    fun groupTitleKey(id: ToolGroupId): String? =
        if (id == ToolGroupId.Home) null else "app.group.${id.id}"

    private fun tool(
        id: ToolId,
        group: ToolGroupId,
        titleKey: String,
        glyph: String,
        keywords: List<String>,
        status: ToolStatus = ToolStatus.Available
    ): ToolDefinition {
        val history = id != ToolId.Mootool && id != ToolId.Hardware && id != ToolId.MessageBoard && id != ToolId.Variables && id != ToolId.Translation
        val favorites = id == ToolId.Regex || id == ToolId.Cron || id == ToolId.ColorBoard
        val navKeywords = buildList {
            addAll(keywords)
            if (id.detachable) {
                add("detach")
                add("分离")
            }
        }
        return ToolDefinition(id, group, titleKey, navKeywords, glyph, status, history, favorites)
    }
}
